ext/date/php_date.c

```diff
@@ -4108,39 +4109,61 @@ static int date_period_initialize(timelib_time **st, timelib_time **et, timelib_
 /* {{{ Creates new DatePeriod object. */
 PHP_METHOD(DatePeriod, __construct)
 {
-	php_period_obj   *dpobj;
-	php_date_obj     *dateobj;
-	zval *start, *end = NULL, *interval;
-	zend_long  recurrences = 0, options = 0;
-	char *isostr = NULL;
-	size_t   isostr_len = 0;
+	php_period_obj *dpobj;
+	php_date_obj *dateobj;
+	zend_object *start_obj;
+	zend_string *start_str;
+	zend_object *interval_obj = NULL;
+	zend_bool interval_is_null = 1;
+	zend_object *end_obj = NULL;
+	zend_long recurrences = 0;
+	zend_bool end_is_null = 1;
+	zend_long options = 0;
+	char *isostr;
+	size_t isostr_len;
 	timelib_time *clone;
 	zend_error_handling error_handling;
 
-	zend_replace_error_handling(EH_THROW, NULL, &error_handling);
-	if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS(), "OOl|l", &start, date_ce_interface, &interval, date_ce_interval, &recurrences, &options) == FAILURE) {
-		if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS(), "OOO|l", &start, date_ce_interface, &interval, date_ce_interval, &end, date_ce_interface, &options) == FAILURE) {
-			if (zend_parse_parameters_ex(ZEND_PARSE_PARAMS_QUIET, ZEND_NUM_ARGS(), "s|l", &isostr, &isostr_len, &options) == FAILURE) {
-				php_error_docref(NULL, E_WARNING, "This constructor accepts either (DateTimeInterface, DateInterval, int) OR (DateTimeInterface, DateInterval, DateTime) OR (string) as arguments.");
-				zend_restore_error_handling(&error_handling);
-				return;
-			}
-		}
-	}
+	ZEND_PARSE_PARAMETERS_START(1, 4)
+		Z_PARAM_STR_OR_OBJ_OF_CLASS(start_str, start_obj, date_ce_interface)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_OBJ_OF_CLASS_OR_LONG_OR_NULL(interval_obj, date_ce_interval, options, interval_is_null)
+		Z_PARAM_OBJ_OF_CLASS_OR_LONG_OR_NULL(end_obj, date_ce_interface, recurrences, end_is_null)
+		Z_PARAM_LONG(options)
+	ZEND_PARSE_PARAMETERS_END();
 
 	dpobj = Z_PHPPERIOD_P(ZEND_THIS);
 	dpobj->current = NULL;
 
-	if (isostr) {
+	if (start_str) {
+		if (interval_obj) {
+			zend_argument_type_error(2, "must be of type ?int when argument #1 ($start) is a string");
+			RETURN_THROWS();
+		}
+
+		if (!end_is_null) {
+			zend_argument_value_error(3, "must be null when argument #1 ($start) is a string");
+			RETURN_THROWS();
+		}
+
+		isostr = ZSTR_VAL(start_str);
+		isostr_len = ZSTR_LEN(start_str);
+
+		zend_replace_error_handling(EH_THROW, NULL, &error_handling);
```

Comment:
Do I see well that automatic warning promotion is only needed for `date_period_initialize` after the other changes?

Reply:
Right